api/src/main/java/org/openmrs/api/db/hibernate/HibernateConceptDAO.java

```diff
@@ -605,75 +622,72 @@ private LinkedHashSet<Concept> transformNamesToConcepts(List<ConceptName> names)
-		query.append(nameQuery);
-		query.append(")^0.2)");
-		
-		List<String> localeQueries = new ArrayList<>();
-		for (Locale locale : locales) {
+	private SearchPredicate newConceptNameQuery(SearchPredicateFactory f, final String name, final boolean searchKeywords, 
+												final Collection<Locale> locales, final boolean searchExactLocale) {
+		return f.bool().with(b -> {
 			if (searchExactLocale) {
-				localeQueries.add(locale.toString());
+				b.must(f.terms().field("locale").matchingAny(locales));
 			} else {
-				String localeQuery = locale.getLanguage() + "* ";
-				if (!StringUtils.isBlank(locale.getCountry())) {
-					localeQuery += " OR " + locale + "^2 ";
-				}
-				localeQueries.add(localeQuery);
+				b.must(f.bool().with(bb -> {
+					for (Locale locale : locales) {
+						bb.should(f.wildcard().field("locale").matching(locale.getLanguage() + "*"));
+						if (!StringUtils.isBlank(locale.getCountry())) {
+							bb.should(f.match().field("locale").matching(locale).boost(2f));
+						}
+					}
+				}));
 			}
-		}
-		query.append(" locale:(");
-		query.append(StringUtils.join(localeQueries, " OR "));
-		query.append(")");
-		query.append(" voided:false");
-		
-		return query.toString();
+			b.filter(f.match().field("voided").matching(false));
+			
+			b.must(f.bool().with(bb -> {
+				String escapedName = SearchQueryUtils.escapeQuery(name);
+				List<String> tokenizedName = tokenizeName(escapedName, locales);
+				String query = newNameQuery(tokenizedName, escapedName, searchKeywords);
+
+				bb.should(f.match().field("concept.conceptMappings.conceptReferenceTerm.code").matching(name).boost(10f));
+				bb.should(f.and()
+					.add(f.simpleQueryString().field("name").matching(query))
+					.add(f.match().field("localePreferred").matching(true)).boost(1f));
+				bb.should(f.simpleQueryString().field("name").matching(query).boost(0.5f));
+			}));
+		}).toPredicate();
 	}
 	
-	private StringBuilder newNameQuery(final List<String> tokenizedName, final String escapedName,
+	private String newNameQuery(final List<String> tokenizedName, final String escapedName,
 	        final boolean searchKeywords) {
 		final StringBuilder query = new StringBuilder();
 		query.append("(");
 		if (searchKeywords) {
 			//Put exact phrase higher
-			query.append(" name:(\"").append(escapedName).append("\")^0.7");
+			query.append("(\"").append(escapedName).append("\")^0.7");
 			
 			if (!tokenizedName.isEmpty()) {
-				query.append(" OR (");
+				query.append(" | (");
 				for (String token : tokenizedName) {
-					query.append(" (name:(");
+					query.append(" +((");
 					
 					//Include exact
 					query.append(token);
-					query.append(")^0.6 OR name:(");
+					query.append(")^0.6 | (");
 					
 					//Include partial
 					query.append(token);
-					query.append("*)^0.3 OR name:(");
+					query.append("*)^0.3 | (");
 					
 					//Include similar
 					query.append(token);
-					query.append("~0.8)^0.1)");
+					query.append("~1,3)^0.1)");
```

Comment:
Similar implementation has changed. It is now set to differ by one sign and at least 3 first signs need to match.